### PR TITLE
[ADD] stock_landed_costs: Add test case for incoming shipment.

### PR DESCRIPTION
**Pull request by atp-odoo**
Task: https://www.odoo.com/web?#id=19693&view_type=form&model=project.task&menu_id=3942&action=327
pad: https://pad.odoo.com/p/r.g8GNmviPc7vFEis3

### CODE REVIEW

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            'name': self.productB.name,
+            'product_id': self.productB.id,
+            'product_uom_qty': 10,
+            'product_uom': self.productB.uom_id.id,
+            'picking_id': picking_in.id,
+            'location_id': self.supplier_location_id,
+            'location_dest_id': self.stock_location_id})
+        # Confirm incoming shipment.
+        picking_in.action_confirm()
+        # Trasfer incoming shipment
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: typo Transfer

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        picking_in.do_transfer()
+        self.assertEqual(picking_in.state, 'done', 'Not Done incoming shipment')
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: improve assert message like : Incoming Shipment should be in done state.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                (0, 0, {'name': 'equal split', 'split_method': 'equal', 'price_unit': value['equal_price_unit'], 'product_id': self.productLC.id}),
+                (0, 0, {'name': 'split by quantity', 'split_method': 'by_quantity', 'price_unit': value['quantity_price_unit'], 'product_id': self.productLC.id}),
+                (0, 0, {'name': 'split by weight', 'split_method': 'by_weight', 'price_unit': value['weight_price_unit'], 'product_id': self.productLC.id}),
+                (0, 0, {'name': 'split by volume', 'split_method': 'by_volume', 'price_unit': value['volume_price_unit'], 'product_id': self.productLC.id})
+            ],
+            valuation_adjustment_lines=[]
+        ))
+
+    def test_00_create_landed_costs(self):
+        """ Products Incomeing shipment & Evaluate Landed Costs """
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: typo: incoming...

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        stock_landed_cost = self._create_landed_costs({'equal_price_unit': 10, 'quantity_price_unit': 150, 'weight_price_unit': 250, 'volume_price_unit': 20}, self._create_picking_in())
+        stock_landed_cost.compute_landed_cost()
+        for valuation in stock_landed_cost.valuation_adjustment_lines:
+            if valuation.cost_line_id.name == 'equal split':
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: don't check with name .. better to compare with split_method.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                self.assertEqual(valuation.additional_landed_cost, 5.0)
+            elif valuation.cost_line_id.name == 'split by quantity' and valuation.move_id.name == "Product A":
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: same here. compare with split_method

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            elif valuation.cost_line_id.name == 'split by weight' and valuation.move_id.name == "Product A":
+                self.assertEqual(valuation.additional_landed_cost, 50.0)
+            elif valuation.cost_line_id.name == 'split by weight' and valuation.move_id.name == "Product B":
+                self.assertEqual(valuation.additional_landed_cost, 200.00)
+            elif valuation.cost_line_id.name == 'split by volume' and valuation.move_id.name == "Product A":
+                self.assertEqual(valuation.additional_landed_cost, 5.0)
+            elif valuation.cost_line_id.name == 'split by volume' and valuation.move_id.name == "Product B":
+                self.assertEqual(valuation.additional_landed_cost, 15.0)
+            else:
+                raise 'Unrecognized Valuation Adjustment Line'
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: cannot use assertRaises ?

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        #Validate the landed cost
+        stock_landed_cost.button_validate()
+        self.assertEqual(stock_landed_cost.state, 'done')
+        self.assertTrue(stock_landed_cost.account_move_id)
+        #Product B - 8, Product A(WH/IN) - 8, Product A(WH/OUT) - 8 (8+8+8=24 - Account journals Items)
+        self.assertEqual(len(stock_landed_cost.account_move_id.line_ids), 16)
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: what happen if assert failed? required message.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+    def test_01_create_landed_costs(self):
+        """ Products Incomeing shipment then Outgoing shipment & Evaluate Landed Costs """
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: again typo. ;)

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        picking_out.action_assign()
+        # Check availability for product A
+        aval_a_qty = self.Move.search([('product_id', '=', self.productA.id), ('picking_id', '=', picking_out.id)], limit=1).availability
+        self.assertEqual(aval_a_qty, 2.0, 'Wrong move quantity availability of product A (%s found instead of 5)' % (aval_a_qty))
+        # Transfer picking.
+        picking_out.do_transfer()
+        # Check available product A
+        aval_a_qty = self.Product.browse(self.productA.id).qty_available
+        self.assertEqual(aval_a_qty, 3.0, 'Wrong available quantity of product A (%s found instead of 3)' % (aval_a_qty))
+        self.assertEqual(picking_out.state, 'done', 'Not Done Outgoing shipment')
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: improve message.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        """ Create Landed Cost for incoming shipment. """
+        stock_landed_cost = self._create_landed_costs({'equal_price_unit': 10, 'quantity_price_unit': 150, 'weight_price_unit': 250, 'volume_price_unit': 20}, picking_in1)
+        stock_landed_cost.compute_landed_cost()
+        for valuation in stock_landed_cost.valuation_adjustment_lines:
+            if valuation.cost_line_id.name == 'equal split':
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: same as above check with split_method..

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            'location_dest_id': self.customer_location_id})
+        self.Move.create({
+            'name': self.productA.name,
+            'product_id': self.productA.id,
+            'product_uom_qty': 2,
+            'product_uom': self.productA.uom_id.id,
+            'picking_id': picking_out.id,
+            'location_id': self.stock_location_id,
+            'location_dest_id': self.customer_location_id})
+        # Confirm outgoing shipmentst.
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: typo shipment.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            elif valuation.cost_line_id.name == 'split by weight' and valuation.move_id.name == "Product A":
+                self.assertEqual(valuation.additional_landed_cost, 50.0)
+            elif valuation.cost_line_id.name == 'split by weight' and valuation.move_id.name == "Product B":
+                self.assertEqual(valuation.additional_landed_cost, 200.00)
+            elif valuation.cost_line_id.name == 'split by volume' and valuation.move_id.name == "Product A":
+                self.assertEqual(valuation.additional_landed_cost, 5.0)
+            elif valuation.cost_line_id.name == 'split by volume' and valuation.move_id.name == "Product B":
+                self.assertEqual(valuation.additional_landed_cost, 15.0)
+            else:
+                raise 'Unrecognized Valuation Adjustment Line'
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: use assertRaises

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+            elif valuation.cost_line_id.name == 'split by weight' and valuation.move_id.name == "Product A":
+                self.assertEqual(valuation.additional_landed_cost, -10.0)
+            elif valuation.cost_line_id.name == 'split by weight' and valuation.move_id.name == "Product B":
+                self.assertEqual(valuation.additional_landed_cost, -40.0)
+            elif valuation.cost_line_id.name == 'split by volume' and valuation.move_id.name == "Product A":
+                self.assertEqual(valuation.additional_landed_cost, -1.25)
+            elif valuation.cost_line_id.name == 'split by volume' and valuation.move_id.name == "Product B":
+                self.assertEqual(valuation.additional_landed_cost, -3.75)
+            else:
+                raise 'Unrecognized Valuation Adjustment Line'
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: same here use assertRaises

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        self.picking_out.action_confirm()
+        #Product assign to outgoing shipments
+        self.picking_out.action_assign()
+        #Check availability for product A.
+        aval_a_qty = self.Move.search([('product_id', '=', self.productA.id), ('picking_id', '=', self.picking_out.id)], limit=1).availability
+        self.assertEqual(aval_a_qty, 2.0, 'Wrong move quantity availability of product A (%s found instead of 2)' % (aval_a_qty))
+        #Transfer picking.
+        self.picking_out.do_transfer()
+        #Check available product A.
+        aval_a_qty = self.Product.browse(self.productA.id).qty_available
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: self.productA is recordset you can directly access self.productA.qty_available.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        self.assertTrue(self._validate_additional_landed_cost_lines(stock_negative_landed_cost, valid_val), 'Unrecognized Valuation Adjustment Line')
+        #Validate the landed cost.
+        stock_negative_landed_cost.button_validate()
+        self.assertEqual(stock_negative_landed_cost.state, 'done')
+        self.assertTrue(stock_negative_landed_cost.account_move_id)
+        #Product B - 8, Product A(WH/IN) - 8, Product A(WH/OUT) - 8 (8+8+8=24 - Account journals Items).
+        self.assertEqual(len(stock_landed_cost.account_move_id.line_ids), 24, ' Landed Cost lines should be 24 instead of %s' % (len(stock_landed_cost.account_move_id.line_ids)))
+        account_landed_entrys = self.env['account.move.line'].search(['|', ('ref', '=', stock_landed_cost.name), ('ref', '=', stock_negative_landed_cost.name)])
+        total_credits = float_round(sum([account_landed_entry.credit for account_landed_entry in account_landed_entrys]), precision_digits=2)
+        total_debits = float_round(sum([account_landed_entry.credit for account_landed_entry in account_landed_entrys]), precision_digits=2)
```
Comment (atp-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: account_landed_entry.debit here . ;)

File: addons/stock_landed_costs/stock_landed_costs.py [hunk 16]
```diff
@@ -267,7 +267,8 @@ def compute_landed_cost(self, cr, uid, ids, context=None):
 
                         if digits:
                             value = float_round(value, precision_digits=digits[1], rounding_method='UP')
-                            value = min(value, line.price_unit - value_split)
+                            fnc = min if line.price_unit > 0 else max
```
Comment (hmo-odoo) [inline addons/stock_landed_costs/stock_landed_costs.py]: why need this changes ? is it bug ?

File: addons/stock_landed_costs/tests/common.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+
+from odoo.tests import common
+
+
+class TestStockLandedCostsCommon(common.TransactionCase):
```
Comment (hmo-odoo) [inline addons/stock_landed_costs/tests/common.py]: need to clean test case with nice example (real use case of landed cost)

File: addons/stock_landed_costs/tests/common.py [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+            'name': 'Microwave Oven',
+            'type': 'product',
+            'cost_method': 'real',
+            'valuation': 'real_time',
+            'standard_price': 1.0,
+            'weight': 20,
+            'volume': 1.5,
+            'categ_id': self.categ_all.id})
+        # Create service type product 1.Labour 2.Brokerage 3.Transportation 4.Packaging
+        self.labour_equal = self._create_services('Labour Cost')
```
Comment (jco-odoo) [inline addons/stock_landed_costs/tests/common.py]: Labour Cost is a little bit weird as landed cost

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+        #         -------------------------------------------
+        #         1.labour            -5        By Equal
+        #         2.brokerage         -50       By Quantity
+        #         3.transportation    -50       By Weight
+        #         4.packaging         -5        By Volume
+
+        # Process incoming shipment
+        income_ship = self._process_incoming_shipment()
+        # Refrigerator outgoing shipment.
+        self._process_outgoign_shipment()
```
Comment (jco-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: outgoing

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+            'by_volume_refrigerator': 5.0,
+            'by_volume_oven': 15.0}
+
+        # Check valuation adjustment line recognized or not
+        self._validate_additional_landed_cost_lines(stock_landed_cost, valid_vals)
+        # Validate the landed cost.
+        stock_landed_cost.button_validate()
+        self.assertTrue(stock_landed_cost.account_move_id, 'Landed costs should be available account move lines')
+        account_entry = self.env['account.move.line'].read_group(
+            [('move_id', '=', stock_landed_cost.account_move_id.id)], ['debit', 'credit', 'move_id'], ['move_id'])[0]
```
Comment (jco-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: Checking that debit and credit are equal, does not test if the amounts are correct.

File: addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+            'by_volume_oven': -3.75}
+        # Check valuation adjustment line recognized or not
+        self._validate_additional_landed_cost_lines(stock_negative_landed_cost, valid_vals)
+        # Validate the landed cost.
+        stock_negative_landed_cost.button_validate()
+        self.assertEqual(stock_negative_landed_cost.state, 'done', 'Negative landed costs should be in done state')
+        self.assertTrue(stock_negative_landed_cost.account_move_id, 'Landed costs should be available account move lines')
+        account_entry = self.env['account.move.line'].read_group(
+            [('move_id', '=', stock_negative_landed_cost.account_move_id.id)], ['debit', 'credit', 'move_id'], ['move_id'])[0]
+        self.assertEqual(account_entry['debit'], account_entry['credit'], 'Wrong Account Entry')
```
Comment (jco-odoo) [inline addons/stock_landed_costs/tests/test_stock_landed_costs_purchase.py]: Should check amounts on account_move_lines here too. (normally also the accounts used)

**Pull request merged**